Baseline resampling

benmoran11 opened the pull request:
infer_mixture now has a new option, method = "BR". If selected, infer_mixture uses a new internal C++ function, gsi_mcmc_fb, which performs MCMC using the Pella-Masuda model, with explicit simulations of reference allele frequencies updated based on mixture fish allocations. This new function is implemented using RcppParallel, such that runtimes are very reasonable on machines with 4+ cores.

Commit:
Fully Bayesian MCMC near complete, but fatal error in package build

Commit:
Fully Bayesian MCMC

Commit:
infer_mixture with full Bayesian added

Commit:
Fold baseline resampling into original infer_mixture function

Commit:
Made MCMC & BR methods independent, and optimizing BR

Commit:
Streamline baseline resampling

Commit:
Remove unnecessary likelihood scaling

Commit:
Revert "Remove unnecessary likelihood scaling"

This reverts commit a5dd5185936fed0a28a59ca59b08e9f4d7141d1e.

Commit:
Revert "Streamline baseline resampling"

This reverts commit 1a1951ac9bc188319376d508b241eb8882fa95ed.

Commit:
Re-reverting aspects of BR MCMC after troubleshooting

Commit:
Streamline loglikelihood sweeping

Commit:
Update Documentation

Commit:
Foundations for RcppParallel integration

Commit:
Simulating allele frequencies (theta) for direct genotype logL calculations, and preliminary MCMC switch for BR method.

benmoran11 committed:
merge parallel foundations with theta simulation

benmoran11 committed:
Functional parallelized MCMC

benmoran11 committed:
Trimmed MCMC function

benmoran11 committed:
Stabilized parallel implementation

benmoran11 committed:
Stabilized parallel implementation

benmoran11 committed:
Cleaning unused functions

eriqande committed:
Added prelim_burn_in option and clarified runtime messages.

The runtime messages made it sound like burn-in sweeps were added at the beginning of data-collection sweeps.  But in fact, the total number of sweeps is fixed, and the data-collection sweeps are all the remain after discarding the burn-in.

eriqande commented:
Ben! This is outstanding!

I fixed a few things in the last commit.  They were not problems that your changes had left, but rather some inaccuracies in the runtime messages I found when I was testing stuff.  I also added the `prelim_burn_in` option and a little error checking on that because I think it is likely that people may want different degrees of burn-in for that initialization run and for the BR runs.

One thing that I think would be great to add would be a vignette named something like
`rubias-fully-bayesian.Rmd` that just runs through:

1. how to use the "BR" option
    - including the `prelim_reps` and `prelim_burn_in` options.  Maybe show if that creates any difference in the results.
2. Comparisons of results between the plain version and the BR version.  Maybe with some plots like the attached one below.
3. We gotta be sure to let people know how to use `RcppParallel::setThreadOptions(numThreads = 1)` to control the number of threads.  This is really important if they are on an HPC system.  In that case RcppParallel might detect all the cores, and try to use them, but if you have only requested 4 cores, say, (or 1) then the queuing system may kill the job for trying to use too many cores.  Also `RcppParallel::setThreadOptions(numThreads = RcppParallel::defaultNumThreads())` to set things back to the default of all cores.

You should be able to copy the formatting/setup for that vignette from `rubias/vignettes/rubias-overview.Rmd`, but make it a lot shorter so that it doesn't take forever to knit it (although having it run through at least a little of the BR stuff is good, as it provides a check for breaking changes in the future).


<img width="909" alt="boing" src="https://user-images.githubusercontent.com/2035238/58764090-5d2a1f00-8520-11e9-95fa-8119e1fecf01.png">

I will pull this in soon, but I saw that Travis was bombing some checks.  Then I went to run R CMD CHECK and had to update roxygen to 6.1.1, and now I can't get anything to build.  Grrr.

So, won't merge this till I fix it.  In the meantime, feel free to work on the vignette on this branch.  Anything you push to it will get added to the pull request.

Great work!

eriqande committed:
Add prelim_burn_in to documentation.

eriqande committed:
Add RcppParallel Libs to NAMESPACE via roxygen

Simple fix to make sure that NAMESPACE gets auto-generated by roxygen2.

This fixes issues Eric was having with building the package.  Yay!

eriqande committed:
Updated the README and the rubias-overview vignette

added short blurbs about the fully-bayesian model and directed the read to the vignette about that.

eriqande commented:
Hey Ben!

I fixed the issues I was having and updated the README and the rubias-overview vignette.  Before pulling this all in, I'll have you make a quick `rubias-fully-bayesian.Rmd` vignette (give it short run times on a small data set so that it doesn't take too long to produce when checking on travis, etc.)

Then, when I am back stateside, we can run the ADFG example data set on the 24 and 20 core Mac and Linux boxes back in Santa Cruz, and then pull this all into master on next Saturday, and then submit to CRAN.  Cool!

eriqande reviewed (approved):
Hey Ben!
This looks super.  I made a few commit and comments.  As you'll see, I think it is looking great.  You'll want to pull in my commits on the baseline_resampling branch.  Then we can pull it all into master when I am back.  It could use a short vignette.  I hope you are getting some great outdoor time in FoCo.

benmoran11 committed:
Vignette for the fully Bayesian model

eriqande committed:
Minor typo fixes.

eriqande committed:
tweak

eriqande committed:
Final additions (NEWS, comments, version) before CRAN submission.
